controller_manager/src/controller_manager.cpp

```diff
@@ -1603,10 +1603,25 @@ void ControllerManager::activate_controllers(
       {
         RCLCPP_ERROR(
           get_logger(),
-          "After activation, controller '%s' is in state '%s' (%d), expected '%s' (%d).",
+          "After activation, controller '%s' is in state '%s' (%d), expected '%s' (%d). Releasing "
+          "interfaces!",
           controller->get_node()->get_name(), new_state.label().c_str(), new_state.id(),
           hardware_interface::lifecycle_state_names::ACTIVE,
           lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+        controller->release_interfaces();
+        // Now prepare and perform the stop interface switching as this is needed for exclusive
+        // interfaces
+        if (
+          !command_interface_names.empty() &&
+          (!resource_manager_->prepare_command_mode_switch({}, command_interface_names) ||
+           !resource_manager_->perform_command_mode_switch({}, command_interface_names)))
```

Comment:
We have here a bit of a chicken-egg problem. Currently we don't switch controller if this command fails (see line 2112), and we react on an error in general and not an error of specific hardware. We will have to change this in the future, but not for now as it seems not be a problem.

Back to the topic...

I thinks that a cleaner soluation would be to gather all the command interfaces of controller that failed to start and then do only once call on this method, to poke arround HW as little as possible. In current implementation we are doing N-time check in M HW regaridng the interface, if we add this to the line 2136 then we will do only 1xM checks.

Reply:
I would do as you said, to trigger for all failed controllers at the end!